### PR TITLE
Patient generation

### PR DESCRIPTION
**Pull request by mayerm94**
Add function to generate CQM Patient(s) that has a qdmPatient that is fully-loaded with all data types which are populated with all possible attributes.

Pull requests into cqm-models require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/CYPRESS-431
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] If there were any JavaScript changes, this PR has updated the `dist` directory using `npm run dist`
- [x] If applicable, the library version number in `package.json` and `cqm-models.gemspec` has been updated
- [x] All changes can be reproduced by running the generator script
- [x] Cqm-execution fixtures have been updated with the update_cqm_execution_fixtures.sh script inside server-scripts using this branch in the cqm-converter

**Bonnie Reviewer:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] You have executed the generator script, and made sure no changes were made that the generator did not reproduce itself

**Cypress Reviewer:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] You have executed the generator script, and made sure no changes were made that the generator did not reproduce itself

### CODE REVIEW

File: app/models/models.rb [hunk 1]
```diff
@@ -12,6 +12,8 @@ module QDM
 require_relative 'qdm/basetypes/interval'
 require_relative 'qdm/basetypes/quantity'
 require_relative 'qdm/basetypes/ratio'
+# FacilityLocation is technically a basetype
+require_relative 'qdm/facility_location'
```
Comment (losborne) [inline app/models/models.rb]: There is a duplicate `require_relative 'qdm/facility_location'` down below added by looping through the datatypes.   Adding `<%- next if datatype.underscore == 'facilityLocation' -%>` to the template file should fix that.

File: lib/generate_patients.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+require_relative '../app/models/models'
+require('generate_types')
+
+module QDM
+  # PatientGeneration module can be used to generate patients with dataElements that are populated
+  # with every possible attribute that said type supports
+  module PatientGeneration
+    # Generates patient(s) with fully-loaded dataElements if patient_per_type is false then a
+    # single patient will be returned that has every data elemetnt on it
+    def self.generate_exhastive_data_element_patients(patient_per_type = true)
```
Comment (losborne) [inline lib/generate_patients.rb]: This function should be 2 separate functions without the `patient_per_type` boolean passed in.  The comments and return logic below are already somewhat separated, so it should be easy to refactor.

File: lib/generate_patients.rb [hunk 3]
```diff
@@ -0,0 +1,132 @@
+require('nokogiri')
+require_relative '../app/models/models'
+require('generate_types')
+
+module QDM
+  # PatientGeneration module can be used to generate patients with dataElements that are populated
+  # with every possible attribute that said type supports
+  module PatientGeneration
+    # Generates patient(s) with fully-loaded dataElements if patient_per_type is false then a
+    # single patient will be returned that has every data elemetnt on it
```
Comment (losborne) [inline lib/generate_patients.rb]: elemetnt -> element

File: lib/generate_patients.rb [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    def self.generate_exhastive_data_element_patients(patient_per_type = true)
+      datatypes = get_datatypes(File.join(File.dirname(__FILE__), '../modelinfo/qdm-modelinfo-5.4.xml'))
```
Comment (losborne) [inline lib/generate_patients.rb]: Instead of hardcoding which version of the modelinfo file this uses, would it be better to pass in the path as an argument?  Not necessarily, but something to consider.

File: lib/generate_patients.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        datatypes.push(datatype_name)
+      end
+      datatypes
+    end
+
+    def self.generate_loaded_datatype(data_element_type, negate_data_element = false)
+      data_element = QDM.const_get(data_element_type).new
+      fields = data_element.typed_attributes.keys
+      fields.each do |field_name|
+        # Ignore these fields, they areused by mongoose
```
Comment (losborne) [inline lib/generate_patients.rb]: areused -> are used

File: lib/generate_patients.rb [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        # Skip the setting of patient description for now
+      elsif field_name == 'negationRationale'
+        data_element[field_name] = QDM::BaseTypeGeneration.generate_code_field if negate_data_element
+      elsif field_name == 'components'
+        data_element[field_name] = [QDM::BaseTypeGeneration.generate_component]
+      elsif field_name == 'result'
+        # TODO: Result can be MANY Integer, Decimal, Code, Quantity or Ratio randomize this
+        data_element[field_name] = QDM::BaseTypeGeneration.generate_code_field
+      elsif field_name == 'diagnoses'
+        # TODO: Randomize counta and contents of diagnosis
```
Comment (losborne) [inline lib/generate_patients.rb]: counta -> count

File: lib/generate_patients.rb [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+      end
+    end
+
+    def self.populate_field_by_type(field_name, data_element)
+      field_type = data_element.fields[field_name].type
+      if field_type == QDM::Code
+        data_element[field_name] = QDM::BaseTypeGeneration.generate_code_field
+      elsif field_type == Date
+        # TODO: Randomize date
+        data_element[field_name] = Date.today
```
Comment (losborne) [inline lib/generate_patients.rb]: This `elsif` block is not hit by tests--is `Date` the correct `field_type`?

File: lib/generate_patients.rb [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+require('generate_types')
+
+module QDM
+  # PatientGeneration module can be used to generate patients with dataElements that are populated
+  # with every possible attribute that said type supports
+  module PatientGeneration
+    # Generates patient(s) with fully-loaded dataElements if new_patient_for_each_type is false then a
+    # single patient will be returned that has every data element on it
+    def self.generate_exhastive_data_element_patients(new_patient_for_each_type = true)
+      datatypes = get_datatypes(File.join(File.dirname(__FILE__), '../modelinfo/qdm-modelinfo-5.4.xml'))
```
[the remark this answers is not in the document]
Comment (mayerm94) [reply lib/generate_patients.rb]: If we passed this in as a parameter, those who use the function (currently only used outside this repo) would need to know the file name of the model info file within cqm-models they want to use. In order to use this in future versions without needing to update the line there would need to be a global variable in the project CURRENT_MODELINFO or something along those lines that would get updated each year.

File: lib/generate_patients.rb [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      end
+      datatypes
+    end
+
+    def self.generate_loaded_datatype(data_element_type, negate_data_element = false)
+      data_element = QDM.const_get(data_element_type).new
+      fields = data_element.typed_attributes.keys
+      fields.each do |field_name|
+        # Ignore these fields, they are used by mongoose
+        next if %w[_id _type].include? field_name
```
Comment (jbradl11) [inline lib/generate_patients.rb]: How would mongoose interact with this? Do you mean mongo?

File: lib/generate_patients.rb [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    def self.populate_fields(field_name, data_element, negate_data_element)
+      # There are certain fields that we want to populate manually
+      if field_name == 'description'
+        # Skip the setting of patient description for now
+      elsif field_name == 'negationRationale'
+        data_element[field_name] = QDM::BaseTypeGeneration.generate_code_field if negate_data_element
+      elsif field_name == 'components'
+        data_element[field_name] = [QDM::BaseTypeGeneration.generate_component]
+      elsif field_name == 'result'
+        # TODO: Result can be MANY Integer, Decimal, Code, Quantity or Ratio randomize this
```
Comment (jbradl11) [inline lib/generate_patients.rb]: Is there a planned task in the future to address these TODOs? If not, perhaps we should remove all of the TODOs and have one more general todo of randomizing the generated data.

Comment (mayerm94) [reply lib/generate_patients.rb]: https://jira.mitre.org/browse/CYPRESS-432
